packages/webui/src/client/ui/RundownView/SelectedElementsContext.tsx

```diff
@@ -0,0 +1,165 @@
+import React from 'react'
+import {
+	AdLibActionId,
+	PartInstanceId,
+	PieceId,
+	RundownId,
+	SegmentId,
+} from '@sofie-automation/corelib/dist/dataModel/Ids'
+
+interface RundownElement {
+	type: 'rundown'
+	id: string
+	elementId: RundownId
+}
+
+interface SegmentElement {
+	type: 'segment'
+	id: string
+	elementId: SegmentId
+}
+
+interface PartInstanceElement {
+	type: 'partInstance'
+	id: string
+	elementId: PartInstanceId
+}
+
+interface PieceElement {
+	type: 'piece'
+	id: string
+	elementId: PieceId
+}
+
+interface AdlibActionElement {
+	type: 'adlibAction'
+	id: string
+	elementId: AdLibActionId
+}
+
+// Union type for all possible elements
+type SelectedElement = RundownElement | SegmentElement | PartInstanceElement | PieceElement | AdlibActionElement
+
+interface SelectionContextType {
+	selectedElements: Map<string, SelectedElement>
+	isSelected: (elementId: string) => boolean
+	clearAndSetSelection: (element: SelectedElement) => void
+	toggleSelection: (element: SelectedElement) => void
+	addSelection: (element: SelectedElement) => void
+	removeSelection: (elementId: string) => void
+	clearSelections: () => void
+	getSelectedCount: () => number
+}
+
+type SelectionAction =
+	| { type: 'CLEAR_AND_SET_SELECTION'; payload: SelectedElement }
+	| { type: 'TOGGLE_SELECTION'; payload: SelectedElement }
+	| { type: 'ADD_SELECTION'; payload: SelectedElement }
+	| { type: 'REMOVE_SELECTION'; payload: string }
+	| { type: 'CLEAR_SELECTIONS' }
+
+// Reducer function
+const selectionReducer = (
+	state: Map<string, SelectedElement>,
+	action: SelectionAction,
+	maxSelections: number
+): Map<string, SelectedElement> => {
+	switch (action.type) {
+		case 'CLEAR_AND_SET_SELECTION': {
+			const newMap = new Map([[action.payload.id, action.payload]])
+			return newMap
+		}
+		case 'TOGGLE_SELECTION': {
+			const next = new Map(state)
+			if (next.has(action.payload.id)) {
+				next.delete(action.payload.id)
+			} else if (next.size < maxSelections) {
+				next.set(action.payload.id, action.payload)
+			}
+			return next
+		}
+		case 'ADD_SELECTION': {
+			if (state.size >= maxSelections) return state
+			const next = new Map(state)
+			next.set(action.payload.id, action.payload)
+			return next
+		}
+		case 'REMOVE_SELECTION': {
+			const next = new Map(state)
+			next.delete(action.payload)
+			return next
+		}
+		case 'CLEAR_SELECTIONS': {
+			return new Map()
+		}
+		default:
+			return state
+	}
+}
+
+export const SelectedElementsContext = React.createContext<SelectionContextType | null>(null)
+
+export const SelectedElementProvider: React.FC<{
+	children: React.ReactNode
+	maxSelections?: number // Optional prop to limit maximum selections
+}> = ({ children, maxSelections = 10 }) => {
+	const [selectedElements, dispatch] = React.useReducer(
+		(state: Map<string, SelectedElement>, action: SelectionAction) => selectionReducer(state, action, maxSelections),
+		new Map()
+	)
+
+	const value = React.useMemo(
+		() => ({
+			selectedElements,
```

Comment:
Does the underlying map need to be exposed? It feels like the user should be using the `isSelected` or `getSelectedCount` functions on the context object. If so, then exposing the `selectedElements` just allows the implementation details to leak. Moving `selectedElements` outside of this useMemo will allow the underlying implementation to be easily changed without any of the users knowing.